src/shared/fdset.c

```diff
@@ -341,8 +427,13 @@ int fdset_iterate(FDSet *s, Iterator *i) {
         const void *k;
         void *v;
 
-        if (!hashmap_iterate(MAKE_HASHMAP(s), i, &v, &k))
-                return -ENOENT;
+        for (;;) {
+                if (!hashmap_iterate(MAKE_HASHMAP(s), i, &v, &k))
+                        return -ENOENT;
+
+                if (PTR_TO_INT(v) >= 0) /* Not an index, done */
```

Comment:
maybe add another helper to make this more readable (and make the comment unnecessary): PTR_IS_FD() or PTR_IS_FD_INDEX().